Sources/SourceKitLSP/Swift/DocumentFormatting.swift

```diff
@@ -0,0 +1,236 @@
+    var range: Range<Int>
+    /// The string the range is being replaced with.
+    var replacement: String
+  }
+
+  var edits: [StringBasedEdit] = []
+  for change in difference {
+    // Adjust the index offset based on changes that `Collection.difference` expects to already have been applied.
+    var adjustment: Int = 0
+    for edit in edits {
+      if edit.range.upperBound < change.offset {
+        adjustment = adjustment + edit.range.count - edit.replacement.count
+      }
+    }
+    let adjustedOffset = change.offset + adjustment
+    let edit =
+      switch change {
+      case .insert(offset: _, element: let element, associatedWith: _):
+        StringBasedEdit(range: adjustedOffset..<adjustedOffset, replacement: String(element))
+      case .remove(offset: _, element: _, associatedWith: _):
+        StringBasedEdit(range: adjustedOffset..<(adjustedOffset + 1), replacement: "")
+      }
+
+    // If we have an existing edit that is adjacent to this one, merge them.
+    // Otherwise, just append them.
+    if let mergableEditIndex = edits.firstIndex(where: {
+      $0.range.upperBound == edit.range.lowerBound || edit.range.upperBound == $0.range.lowerBound
+    }) {
+      let mergableEdit = edits[mergableEditIndex]
+      if mergableEdit.range.upperBound == edit.range.lowerBound {
+        edits[mergableEditIndex] = StringBasedEdit(
+          range: mergableEdit.range.lowerBound..<edit.range.upperBound,
+          replacement: mergableEdit.replacement + edit.replacement
+        )
+      } else {
+        precondition(edit.range.upperBound == mergableEdit.range.lowerBound)
+        edits[mergableEditIndex] = StringBasedEdit(
+          range: edit.range.lowerBound..<mergableEdit.range.upperBound,
+          replacement: edit.replacement + mergableEdit.replacement
+        )
+      }
+    } else {
+      edits.append(edit)
+    }
+  }
+
+  // Map the string-based edits to line-column based edits to be consumed by LSP
+
+  return edits.map { edit in
+    let (startLine, startColumn) = original.lineTable.lineAndUTF16ColumnOf(
+      original.text.index(original.text.startIndex, offsetBy: edit.range.lowerBound)
+    )
+    let (endLine, endColumn) = original.lineTable.lineAndUTF16ColumnOf(
+      original.text.index(original.text.startIndex, offsetBy: edit.range.upperBound)
+    )
+
+    return TextEdit(
+      range: Position(line: startLine, utf16index: startColumn)..<Position(line: endLine, utf16index: endColumn),
+      newText: edit.replacement
+    )
+  }
+}
+
+extension SwiftLanguageServer {
+  public func documentFormatting(_ req: DocumentFormattingRequest) async throws -> [TextEdit]? {
+    let snapshot = try documentManager.latestSnapshot(req.textDocument.uri)
+
+    guard let swiftFormat else {
+      throw ResponseError.unknown("Formatting not supported because toolchain is missing the swift-format executable")
```

Comment:
```suggestion
      throw ResponseError.unknown("Formatting not supported because the toolchain is missing the swift-format executable")
```
Should we include the toolchain path here too 🤔?

Reply:
No, I think that’s not necessary. It’s the toolchain that’s being used for all operations.